File: citizen/consumer/consumer.go
package consumer

import "math"

// ConsumerEngine translates aggregate disposable income and citizen sentiment
// into sector-level consumption demand factors that drive corporate revenue.
type ConsumerEngine struct {
	TotalConsumerSpending float64            // Total annual personal consumption expenditures (PCE) in USD (~$19 Trillion)
	SectorDemandFactors   map[string]float64 // Sector name -> demand multiplier (1.0 = baseline)
}

// NewDefaultConsumerEngine initializes consumer demand at US macroeconomic baseline.
func NewDefaultConsumerEngine() *ConsumerEngine {
	factors := map[string]float64{
		"Information Technology": 1.0,
		"Health Care":            1.0,
		"Financials":             1.0,
		"Consumer Discretionary": 1.0,
		"Consumer Staples":       1.0,
		"Industrials":            1.0,
		"Energy":                 1.0,
		"Materials":              1.0,
		"Communication Services": 1.0,
		"Utilities":              1.0,
		"Real Estate":            1.0,
	}

	return &ConsumerEngine{
		TotalConsumerSpending: 18_500_000_000_000.0, // ~$18.5T US PCE baseline
		SectorDemandFactors:   factors,
	}
}

// Tick calculates sector demand multipliers based on disposable income and spending propensity.
// - disposableIncome: total aggregate after-tax income
// - spendingPropensity: sentiment multiplier (>1 optimistic, <1 conservative)
// - savingsRate: fraction diverted to personal savings
func (c *ConsumerEngine) Tick(disposableIncome, spendingPropensity, savingsRate float64) {
	effectiveSavings := math.Max(0.02, math.Min(0.15, savingsRate/spendingPropensity))
	c.TotalConsumerSpending = disposableIncome * (1.0 - effectiveSavings) * spendingPropensity

	// Elasticity profiles:
	// Highly elastic sectors boom when propensity > 1.0 and pull back when propensity < 1.0.
	// Inelastic sectors remain stable regardless of sentiment.
	propensityDelta := spendingPropensity - 1.0

	c.SectorDemandFactors["Consumer Staples"] = 1.0 + propensityDelta*0.15      // Inelastic (food, household)
	c.SectorDemandFactors["Health Care"] = 1.0 + propensityDelta*0.10            // Highly inelastic (medicine, care)
	c.SectorDemandFactors["Utilities"] = 1.0 + propensityDelta*0.08              // Inelastic (power, water)
	c.SectorDemandFactors["Energy"] = 1.0 + propensityDelta*0.35                 // Moderate (gasoline, heat)
	c.SectorDemandFactors["Communication Services"] = 1.0 + propensityDelta*0.40 // Telecom, media
	c.SectorDemandFactors["Financials"] = 1.0 + propensityDelta*0.50             // Credit, banking
	c.SectorDemandFactors["Materials"] = 1.0 + propensityDelta*0.60              // Raw goods
	c.SectorDemandFactors["Industrials"] = 1.0 + propensityDelta*0.75            // Capital goods
	c.SectorDemandFactors["Real Estate"] = 1.0 + propensityDelta*0.85            // Housing demand
	c.SectorDemandFactors["Information Technology"] = 1.0 + propensityDelta*1.10 // High elasticity (tech upgrades)
	c.SectorDemandFactors["Consumer Discretionary"] = 1.0 + propensityDelta*1.40 // Extreme elasticity (luxury, travel)

	// Clamp to realistic bounds [0.50, 1.80]
	for k, v := range c.SectorDemandFactors {
		c.SectorDemandFactors[k] = math.Max(0.50, math.Min(1.80, v))
	}
}

// DemandForSector returns the consumption demand multiplier for a sector.
func (c *ConsumerEngine) DemandForSector(sector string) float64 {
	if factor, ok := c.SectorDemandFactors[sector]; ok {
		return factor
	}
	return 1.0
}

File: src/citizen/consumer.rs
use std::collections::HashMap;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConsumerEngine {
    pub total_consumer_spending: f64,
    pub sector_demand_factors: HashMap<String, f64>,
}

impl Default for ConsumerEngine {
    fn default() -> Self {
        let mut factors = HashMap::new();
        factors.insert("Information Technology".to_string(), 1.0);
        factors.insert("Health Care".to_string(), 1.0);
        factors.insert("Financials".to_string(), 1.0);
        factors.insert("Consumer Discretionary".to_string(), 1.0);
        factors.insert("Consumer Staples".to_string(), 1.0);
        factors.insert("Industrials".to_string(), 1.0);
        factors.insert("Energy".to_string(), 1.0);
        factors.insert("Materials".to_string(), 1.0);
        factors.insert("Communication Services".to_string(), 1.0);
        factors.insert("Utilities".to_string(), 1.0);
        factors.insert("Real Estate".to_string(), 1.0);

        Self {
            total_consumer_spending: 32_000_000_000.0, // $32B annual consumer spending for developing nation
            sector_demand_factors: factors,
        }
    }
}

impl ConsumerEngine {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn tick(&mut self, disposable_income: f64, spending_propensity: f64, savings_rate: f64) {
        let effective_savings = (savings_rate / spending_propensity).clamp(0.02, 0.15);
        self.total_consumer_spending = disposable_income * (1.0 - effective_savings) * spending_propensity;

        let delta = spending_propensity - 1.0;

        self.sector_demand_factors.insert("Consumer Staples".to_string(), 1.0 + delta * 0.15);
        self.sector_demand_factors.insert("Health Care".to_string(), 1.0 + delta * 0.10);
        self.sector_demand_factors.insert("Utilities".to_string(), 1.0 + delta * 0.08);
        self.sector_demand_factors.insert("Energy".to_string(), 1.0 + delta * 0.35);
        self.sector_demand_factors.insert("Communication Services".to_string(), 1.0 + delta * 0.40);
        self.sector_demand_factors.insert("Financials".to_string(), 1.0 + delta * 0.50);
        self.sector_demand_factors.insert("Materials".to_string(), 1.0 + delta * 0.60);
        self.sector_demand_factors.insert("Industrials".to_string(), 1.0 + delta * 0.75);
        self.sector_demand_factors.insert("Real Estate".to_string(), 1.0 + delta * 0.85);
        self.sector_demand_factors.insert("Information Technology".to_string(), 1.0 + delta * 1.10);
        self.sector_demand_factors.insert("Consumer Discretionary".to_string(), 1.0 + delta * 1.40);

        for val in self.sector_demand_factors.values_mut() {
            *val = val.clamp(0.50, 1.80);
        }
    }

    pub fn demand_for_sector(&self, sector: &str) -> f64 {
        self.sector_demand_factors.get(sector).copied().unwrap_or(1.0)
    }
}
